### PR TITLE
Add Haiku support.

### PR DESCRIPTION
**Pull request by korli**
Build support for Haiku is missing at the moment.

This patch updates the build scripts to build on Haiku, this includes x86 and x86_64.
It apply directly to master. Please consider applying.

PS: this is the follow up to the ticket [RT2392](https://rt.openssl.org/Ticket/Display.html?id=2392)

### CODE REVIEW

File: crypto/include/internal/cryptlib.h [hunk 1]
```diff
@@ -92,7 +92,11 @@ DEFINE_LHASH_OF(MEM);
 # ifndef OPENSSL_SYS_VMS
 #  define X509_CERT_AREA          OPENSSLDIR
 #  define X509_CERT_DIR           OPENSSLDIR "/certs"
-#  define X509_CERT_FILE          OPENSSLDIR "/cert.pem"
+#  ifdef OPENSSL_SYS_HAIKU
```
Comment (richsalz) [inline crypto/include/internal/cryptlib.h]: This kind of gratuitous change is why I dislike supporting random systems.

Comment (korli) [inline crypto/include/internal/cryptlib.h]: Well, I can understand this to be unnecessary. Removed.
Thanks for reviewing.

File: Configurations/10-main.conf [hunk 2]
```diff
@@ -1522,5 +1522,32 @@
...
+    "haiku-common" => {
+        template         => 1,
+        cc               => "cc",
+        cflags           => "",
+        debug_cflags     => "-g -O0",
+        release_cflags   => "-O2 -fomit-frame-pointer ",
+        thread_cflag     => "-D_REENTRANT",
+        sys_id           => "HAIKU",
+        lflags           => "-lnetwork",
+        bn_ops           => "SIXTY_FOUR_BIT_LONG RC4_CHAR",
```
Comment (kroeckx) [inline Configurations/10-main.conf]: A long is always 64 bit, even in 32 bit?

File: Configurations/10-main.conf [hunk 3]
```diff
@@ -1522,5 +1522,32 @@
         shared_extension => ".so.\$(SHLIB_MAJOR).\$(SHLIB_MINOR)",
         ranlib           => "$ENV{'RANLIB'}",
     },
+    "haiku-common" => {
+        template         => 1,
+        cc               => "cc",
+        cflags           => "",
```
Comment (kroeckx) [inline Configurations/10-main.conf]: This might need -DL_ENDIAN?

File: Configurations/10-main.conf (hunk 2 above)
Comment (korli) [inline Configurations/10-main.conf]: Moved to haiku-x86_64.

File: Configurations/10-main.conf (hunk 3 above)
Comment (korli) [inline Configurations/10-main.conf]: -DL_ENDIAN and -Wall are now factorized in haiku-common.
Thanks for the review.

**Comment by korli** [comment]
The PR is rebased following master changes. Any chance for review? Would be nice to get this in 1.1.x.

**Comment by richsalz** [comment]
+1 if another team member reviews, i'll pull it in.

**Comment by korli** [comment]
Can this PR go in 1.1.x? To know if rebasing is worthwhile.

**Comment by t-j-h** [comment]
@richsalz you have another +1 from me

**Comment by levitte** [comment]
+1

**Comment by richsalz** [comment]
please rebase and update; it can go in if you are able to get it done today.

**Comment by korli** [comment]
@richsalz this is rebased and updated with changes in Configurations/10-main.conf. Thanks.

**Comment by richsalz** [comment]
thanks for the rapid update.  pushed to master.

**Comment by dot-asm** [comment]
Sorry to crash the party, but x86_64 build should crash in RC4, because RC4_CHAR is incompatible with assembly modules. If it doesn't crash, then it's purely circumstantial. It can vary from processor type to processor type, but even in favourable case crash would be matter of time in real-life cases. In my opinion it should be reverted...

**Comment by richsalz** [comment]
So does Haiku need no-asm?  I don't have this OS to test.

**Comment by dot-asm** [comment]
> So does Haiku need no-asm?

I don't know what it needs. All I'm saying is that RC4_CHAR configuration option is crash-incompatible with x86_64 RC4 assembly (as well as x86 if you want to know). I'm also saying that this indicates that suggested modifications were not rigorously tested if tested at all. Consequently I'm saying that not even minimal sanity requirement is fulfilled and inclusion is unjustified.

>  I don't have this OS to test.

Neither do I. What I've talked about so far was OS-independent. As far as OS-dependent stuff goes, we should rather think of a way for niche OS maintainers to inject exotic things without modifications to main source, so that they can maintain their platforms themselves. Configurations/* is already ready for this. I mean niche OS we don't have access to should be a separate file. As for headers "entropy", probably there should be a configuration option that allows to specify header file of maintainer's choice...

**Comment by levitte** [comment]
That's a thought! Perhaps we could split up `e_os.h` into something sane and use a config setting to point at an appropriate one. 

However, that won't happen in 1.1.0...

**Comment by korli** [comment]
Haiku openssl 1.1.x is configured with: `config zlib shared no-asm`.
Though rc4test output on x86_64 doesn't indicate any problems (without "no-asm" too).

I see now that 1.0.x/x86_64 only had RC4_CHUNK, which happens to have been removed in 1.1.x. Having RC4_CHAR on 1.1.x/x86_64 is probably a mistake when porting the config from "Configure" 1.0.x. I'd happy to take advice on how to improve the situation - I understand the option RC4_CHAR can be safely removed from the x86_64 config. I can surely check and follow up on this. Sorry for the trouble.

**Comment by richsalz** [comment]
Please open a ne PR that removes rc4_char and add a comment to the config stanza that says “haiku builds with no-asm”
Thanks.

**Comment by korli** [comment]
Please have a look at #969. Thanks.

**Comment by dot-asm** [comment]
Let me rephrase. I have to protest. Once again, if a configuration is not periodically tested by any of the team members, then it doesn't belong in Configurations/*-main.cf. [Some would even argue that it doesn't belong in official tree.] I also object adding assembler modules to configuration and commenting that it doesn't work. If it doesn't work, then it shouldn't be added to start with. I also challenge the need to modify e_os.h in this particular case. As far as I understand Haiku is equipped with GNU compiler, and it supports -include option, which should allow to resolve the problem with configuration file *alone*. That's how it should be done in ideal case, i.e. with a *single* add-on file, and I see no reason why it can't be done in *this* case.

**Comment by richsalz** [comment]
@dot-asm , you are right.  It should not have been in main.  I am changed my mind and am actually inclined to open an MR to revert this.
